### DISCUSSION TITLE
How to build query with Group by Rollup or Cube?

### DISCUSSION THREAD

**Discussion by fullsailor** [juanluispaz/ts-sql-query#58] [Q&A] [answered]
How can I build a `ValueSource` that I can pass to `groupBy()` that can generate a `GROUP BY ROLLUP (col1, col2)` or `GROUP BY CUBE  (col1, col2)`

The syntax defined by Postgres is [here](https://www.postgresql.org/docs/15/sql-select.html). Excerpt:
```
...
 [ GROUP BY [ ALL | DISTINCT ] grouping_element [, ...] ]
...
and grouping_element can be one of:
    ( )
    expression
    ( expression [, ...] )
    ROLLUP ( { expression | ( expression [, ...] ) } [, ...] )
    CUBE ( { expression | ( expression [, ...] ) } [, ...] )
    GROUPING SETS ( grouping_element [, ...] )
```
but other databases implement this query syntax too. ([SQL Server](https://learn.microsoft.com/en-us/sql/t-sql/queries/select-group-by-transact-sql?view=sql-server-ver16))

Thanks you!

**Comment by juanluispaz** [comment, marked as answer]
Hi,

In that case you want to use a Sql expression (like a function call) that is not already mapped in the library, for this case you have the [Sql fragments](https://ts-sql-query.readthedocs.io/en/latest/queries/sql-fragments/).

Hoy can put the Sql fragment in your query or create a reusable function in the DBConnection that generate that fragment for you; and use it as a value in your group by.

Let me know if that works for you.
